crates/oxc_query/src/properties.rs

```diff
@@ -0,0 +1,584 @@
+
+                parts.join(".")
+            }
+        }
+        .into()
+    })
+}
+
+pub(super) fn resolve_assignment_type_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "assignment_to_variable_name" => resolve_property_with(contexts, |v| {
+            let Vertex::AssignmentType(BindingPatternKind::BindingIdentifier(ident)) = v else {
+                return FieldValue::Null;
+            };
+            ident.name.to_string().into()
+        }),
+        _ => {
+            unreachable!(
+                "attempted to read unexpected property '{property_name}' on type 'AssignmentType'"
+            )
+        }
+    }
+}
+
+pub(super) fn resolve_class_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "extended_class_name" => resolve_property_with(contexts, |v| {
+            let Some(Expression::Identifier(ref ident)) = v
+                .as_class()
+                .unwrap_or_else(|| panic!("expected to have a class vertex, instead have: {v:#?}"))
+                .class
+                .super_class
+            else {
+                return FieldValue::Null;
+            };
+            ident.name.to_string().into()
+        }),
+        "is_abstract" => resolve_property_with(contexts, |v| {
+            v.as_class()
+                .unwrap_or_else(|| panic!("expected to have a class vertex, instead have: {v:#?}"))
+                .class
+                .modifiers
+                .contains(oxc_ast::ast::ModifierKind::Abstract)
+                .into()
+        }),
+        _ => {
+            unreachable!("attempted to read unexpected property '{property_name}' on type 'Class'")
+        }
+    }
+}
+
+pub(super) fn resolve_class_method_property<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    property_name: &str,
+    _resolve_info: &ResolveInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, FieldValue> {
+    match property_name {
+        "accessibility" => resolve_property_with(contexts, |v| {
+            v.as_class_method()
+                .unwrap_or_else(|| {
+                    panic!("expected to have a classmethod vertex, instead have: {v:#?}")
+                })
+                .method
+                .accessibility
+                .map(accessibility_to_string)
+                .map_or(FieldValue::Null, Into::into)
+        }),
+        "is_abstract" => resolve_property_with(contexts, |v| {
+            v.as_class_method()
+                .unwrap_or_else(|| {
+                    panic!("expected to have a classmethod vertex, instead have: {v:#?}")
+                })
+                .is_abstract
+                .into()
+        }),
```

Comment:
There are some nice helpers you could use to simplify this kind of code:
```suggestion
        "is_abstract" => resolve_property_with(contexts, field_property!(as_class_method, is_abstract)),
```

The macro expands to the exact same thing you've written here, modulo a slightly different error message.

There's also a `accessor_property!()` macro in case the data is not in a field but behind an accessor function.

Reply:
done in [4f4b165](https://github.com/web-infra-dev/oxc/pull/599/commits/4f4b165ec5add96b3daf5a4a05f25680c4fe186a)